### ISSUE TITLE
D3.1: Virtual images and containers

### ISSUE THREAD

**Issue by minrk** [OpenDreamKit/OpenDreamKit#58]
- **WP3:** [Component Architecture](https://github.com/OpenDreamKit/OpenDreamKit/tree/master/WP3)
- **Lead Institution:** Université de Versailles Saint-Quentin
- **Due:** 2016-02-29 (month 6)
- **Nature:** Other

Creation and distribution of preconfigured cloud oriented virtual machines/containers (e.g. Docker images) for PARI/GP, Sage, SMC, ... See page 43 of the [proposal](https://github.com/OpenDreamKit/OpenDreamKit/raw/master/Proposal/proposal-www.pdf).

We have created virtual images and containers for most of the relevant components of OpenDreamKit. The following projects have been dealt separately:

- MathHub, (containers hosted at https://hub.docker.com/r/kwarc/)
- GAP (containers hosted at https://hub.docker.com/u/gapsystem/)
- OOMMF (virtual image hosted at https://github.com/fangohr/virtualmicromagnetics)
- SageMath, PARI/GP, Singular, etc. (containers hosted at https://hub.docker.com/u/sagemath/)

We report below on each of these components

## MathHub

The MathHub team at Jacobs University maintains two Docker containers for MathHub:

- [MathHub](https://hub.docker.com/r/kwarc/mathhub/) (Dockerfile source https://github.com/KWARC/mathhub_docker)
- [LocalMH](https://hub.docker.com/r/kwarc/localmh/) (Dockerfile source https://hub.docker.com/r/kwarc/localmh/)

The first container is for a MathHub instance, like the one at https://mathhub.info/. The second container is for the LocalMH tool, for offline authoring of MathHub installations.

Both containers are manually built. Build automation via Docker Hub's [automated build service](https://docs.docker.com/docker-hub/builds/) is a potential improvement that will be studied in a next step.

## GAP system

The GAP team maintains two Docker containers:

- [gap-container](https://hub.docker.com/r/gapsystem/gap-container/) (Automated build from https://github.com/gap-system/gap-container).
- [gap-docker](https://hub.docker.com/r/gapsystem/gap-docker/) (Automated build from https://github.com/gap-system/gap-docker).

The first is a minimalistic build containing exclusively the core GAP functionality, without any optional GAP packages (user-contributed extensions, redistributed with GAP). The second is a full GAP install, containing all packages redistributed with GAP. Because some of the packages have non-trivial dependencies on third-party software, gap-docker depends on a [gap-docker-base container](https://hub.docker.com/r/gapsystem/gap-docker-base/) which provides these dependencies. gap-docker-base is not built automatically because of the time and memory restrictions of DockerHub automated builds.

The GAP Docker container is suggested as one of the alternative ways to install GAP: http://www.gap-system.org/Download/alternatives.html.

As a demo application to demonstrate the use of GAP Docker container in the cloud, we run an SCSCP server for the number of groups of order $n$. This service uses GAP Docker container running on a Microsoft Azure virtual machine. Usage details could be found at https://github.com/alex-konovalov/gnu.

GAP is also bundled in the SageMath containers described below.

## OOMMF

The [Virtual Micromagnetics]( http://fangohr.github.io/virtualmicromagnetics/) package is designed to be run through Vagrant provisioned by VirtualBox. The virtual image is hosted at https://atlas.hashicorp.com/virtualmicromagnetics/boxes/full.

The Virtual Micromagnetics developers are considering Docker as an alternative provider for the OOMMF component.

## SageMath

SageMath bundles most of OpenDreamKit components. The community maintains several Docker containers for use and development of SageMath at https://hub.docker.com/u/sagemath/. We describe below the work that has been done in OpenDreamKit regarding those containers.

For SageMath we have created a repository for issue tracking and development of build scripts for SageMath docker images at [sagemath/docker-images](https://github.com/sagemath/docker-images ) (not to be confused with [sagemath/docker](https://github.com/sagemath/docker/) which we have agreed will be superseded by the former).  The repository currently includes build recipes for four Docker images:
 * sagemath/sagemath - This will always provide a build of the most recent released version of SageMath (or previous versions through the use of image tags).  

 * sagemath/sagemath-develop - This provides a build of the most recent develop branch of SageMath from the git repository at the time of building.  This image will need to be updated automatically on a regular basis via an automated mechanism (more on that later).

 * sagemath/sagemath-jupyter - This is the same as the sagemath/sagemath image, but automatically starts the Jupyter notebook with sage when it is run (the base sagemath/sagemath image drops into the sage command line by default).

 * sagemath/sagemath-patchbot - This is the same as the sagemath/sagemath-develop image, but automatically runs the sage patchbot by default. There are some currently unresolved (but likely minor) issues with running the patchbot in this container.

We have found Docker containers to be an effective way to distribute a consistent installation of Sage across multiple platforms--Docker support Linux, Mac OSX, and Windows (with some caveats; more on that below).  Running Sage in a Docker container provides a consistent experience to the user, regardless what the host OS is, and is generally lighter-weight and more transparent to the user than a full VM.  When using the sage command line interface (or any other command line program running in the container) the user interacts with it directly through their host OS's terminal emulator.  The container can also run headless, such as when running the notebook, in which case the user can connect their web browser to the notebook server as though running directly on localhost.

We are also exploring the use of Docker on Windows as a way to provide a simple one-click installer for SageMath on Windows as near-term solution to #66 until and unless Sage can be built and run natively on Windows.  This in turn has driven improvements to the sagemath Docker images.  The installer for Sage would install Docker and its dependencies (if not already installed), and would include the sagemath/sagemath image (or optionally the -develop version) and launcher shortcuts for starting the command-line and notebook interfaces. This installer is being developed in [embray/sage-windows](https://github.com/embray/sage-windows) and will have a prototype ready soon.

The current Docker images for SageMath are based on an Ubuntu 15.10 (Wily Werewolf) base image, though it would be possible to reuse most of the existing build scripts to build SageMath images based on other Linux distributions. This could be useful for Sage development, but less interesting to end-users (see sagemath/docker-images#12).

The Docker Hub service for hosting Docker images does provide an [Automated Build service](https://docs.docker.com/docker-hub/builds/) that could in principle be used to automatically build new sagemath/sagemath-develop images (as well as sagemath/sagemath when there is a new release).  It can monitor changes to the repository storing the image build recipes, and could also be configured to re-build an image upon push to another repository (such as sagemath itself).  Leveraging this service would free us of the overhead of maintaining our own build infrastructure for images.  However, the Docker Hub Automated Build service does have resource restrictions that may prove too limiting for building SageMath:

 * 1 CPU
 * 2 GB of RAM
 * 30 GB disk space
 * 2 hour limit to build time

The disk space is enough, and the RAM is also enough given only 1 CPU is provided.  However with only one CPU it may be difficult to get the build time down to 2 hours, but we have not tried this yet. I (@embray) believe there are several actions that can be taken to improve build time of Sage. In particular, a good many of its dependencies are available, with the correct versions, pre-built from Ubuntu's APT repository (excluding packages that require special patches from Sage that affect runtime behavior).  Improving the Sage build process to allow more reliance of system packages where possible will significantly cut down build time and be beneficial well beyond the case of building Docker images. 

### Caveats:
* Disk space: The Docker images for SageMath are quite large. Originally the sagemath/sagemath image weighed in at 7.361 GB.  It was possible to reduce this to 4.851 GB by cleaning up artifacts of the sage build process at the end of the image build.  For the sagemath/sagemath-develop image we decided to leave those artifacts intact, so running this image gives access to a sage source tree as it would appear immediately after running `make`.  It's possible a few other reductions can be found (such as uninstalling build dependencies once the build is complete), but there's no escaping the fact that these Docker images will be a large download.  Fortunately, Docker images can be exported to a `tar` archive (and further compressed with the compression method of one's choice), for distribution on USB drives and the like.

* Connecting to a notebook or other network interface running in the container does, however, require manual setup of port forwarding when running the Docker container.  The command line option for this is relatively simple to communicate, however, and wrapper scripts / shortcuts may also be provided to users.

* Although the use of Docker containers provides a more or less uniform experience across host OS's, Docker does not technically run natively on Windows or OSX, as the way it works relies heavily on features specific to the Linux kernel. As such, running Docker on Windows or OSX does actually require running a Linux VM in the first place.  The [Docker Toolbox](https://www.docker.com/products/docker-toolbox), which installs Docker on Windows and OSX, also installs VirtualBox and a very small Linux VM, designed specially for Docker, which includes the bare minimum to run Docker and not much else.  Because of its specialized purpose, this VM adds very little overhead compared to a more general purpose VM.

  Docker Toolbox includes a program that makes management of this VM easy ([docker-machine](https://docs.docker.com/machine/)), and because the VM runs headless users will never see it.  The shortcuts installed by Docker Toolbox run a script that ensures the "boot2docker" VM is running, and thus allowing all other `docker` commands to work.  So with a few exceptions users never need to be aware of this underlying virtualization layer:

   * With Docker it is relatively easy to mount existing directories on the host machines as "volumes" accessible from the Docker container (see ["Manage data in containers"](https://docs.docker.com/engine/userguide/containers/dockervolumes/)).  There are a few downsides to this:

     * Although any number of directories can be mounted, they can only be mounted when a container is first created, so one has to think ahead and ensure that all data one will need to access within the container will be accessible (if in doubt, one could mount their entire filesystem I suppose).  In principle it is possible to add new mounts from within a container, but not in a way that is friendly to novices (requires using the `mount` command).

     * On Windows and OSX, because of the two layers of virtualization, any local directory that one wants to mount within a Docker container needs to also first be mounted in the boot2docker VM that Docker itself runs in, so it's a two step process and probably hard to explain to novices.  On the bright side, boot2docker *automatically* mounts the user's home directory, so this is immediately available.  And in many cases all the files a user cares about should be in their home directory, but that is of course often not the case as well.  Another possible "bright side" is that VirtualBox includes a command-line interface for controlling VMs--[`VBoxManage`](https://www.virtualbox.org/manual/ch08.html).  Although not novice-friendly, this does enable writing novice-friendly interfaces for mounting additional directories if they need to, and I (@embray) am working on such interfaces in the Windows installer for Sage (and maybe, hopefully, upstream to Docker as well).

   * Networking has the same problem that directory mounting has--it is fairly easy, when starting a new Docker container, to set up which TCP ports should be forwarded from the Docker container to ports on the host machine.  However, this is harder to do after a container has already been started, and one might have to start a new container if they don't forward the ports properly.  Providing users with shortcuts that do this automatically helps.  Windows and OSX have the same problem with ports needing to be forwarded on the boot2docker VM as well, though this can also be managed with `VBoxManage`.

   * Probably the biggest hurdle to setting up Docker on Windows and OSX is the requirement for [Hardware Assisted Virtualization](https://en.wikipedia.org/wiki/Hardware-assisted_virtualization) (HAV) to be enabled on the user's system.  This is not so much a limitation of Docker as it is a limitation of VirtualBox--it can't run the boot2docker VM (or any 64-bit VM) without HAV.  The default HAV settings vary from system to system, though it is often disabled by default as a security measure.  Some systems provide a tool to change this and other BIOS settings from within Windows, but more likely users will have to change this setting from their BIOS configuration menu.  This is a support nightmare since so many BIOS menus are different and there is no standard.  That said, most BIOS menus have improved substantially in the last decade, so I think helping users with this is less hassle than some other options.  But it is certainly discouraging to tell novices "Now to install Sage you have to reboot your computer, hold down ESC, or DEL, or F10, or Ctrl-C, or Fn-Ctrl-Break, etc. etc. and then rummage around in this plain text curses interface until you find something that says something something about "virtualization" or maybe "VT-X" and make sure that's enabled.  Then press F10 or F8 or S or etc. etc. to save the settings and reboot. Make sure not to change the time!"

     I think it's worth trying this out in at least one workshop to see how bad it is in practice though.  The broader Docker community may also have some useful experience with this, as this is currently the *only* way to get Docker on Windows.

     A possible workaround is to provide a 32-bit version of boot2docker, which is theoretically possible (though it will run slower *without* HAV).  But this is not officially supported by the Docker project, and isn't likely to be (docker/docker#136).

**Comment by sebasguts** [comment]
Some ideas from the Kickoff-Meeting can be found here:
http://opendreamkit.org/meetings/2015-09-02-Kickoff/docker/

I have created an organization on GitHub to store repositories: https://hub.docker.com/u/opendreamkit 
Please send me an email with your Docker Hub account name, so I can add you, if you want.

Initial snippets for Dockerfiles will follow soon.

**Comment by alex-konovalov** [comment]
The DockerHub name for GAP is `gapsystem`. You may see GAP account at https://hub.docker.com/u/gapsystem/.

**Comment by slel** [comment]
For SageMath there is https://hub.docker.com/r/sagemath/sage/

**Comment by sebasguts** [comment]
Since the ideas page was updated, maybe we can start a discussion here. The idea producing a private ppa with source packages sounds good. Then the dockerfiles will be very short and clean. Maybe we can start by collecting some packages?

**Comment by fangohr** [comment]
Hi all, it would be good to also have docker images and ppa with source images for OOMMF (the micromagnetic package we are using in one of the demonstrators for OpenDreamKit) and maybe other micromagnetic tools. We have some work ongoing on this in Southampton, led by Mark Vousden @mvousden . Can I suggest that he joins the effort here and we try to bring everything together in the same framework?

**Comment by sebasguts** [comment]
Maybe it would be a doof thing to collect the stuff, which is there. Singular has a ppa, see http://www.singular.uni-kl.de/index.php/singular-download/110.html .
Polymake will soon (15.10) be installable from the official ubuntu repositories.

**Comment by nthiery** [comment]
@sebasguts: +1 on starting by collecting some packages as you suggest
@fangohr: +1 as well on having @mvousden join.
cc @embray

**Comment by fangohr** [comment]
How do people feel about conda? Seems to be up-and-coming and is packaging more than just Python.

**Comment by slel** [comment]
For SageMathCell, the Sage Cell Server, see
https://groups.google.com/d/msg/sage-support/4xSTjIbwbjo/ppaahx_SCgAJ

**Comment by defeo** [comment]
@kohlhase can you link here pointers to Mathub.info containers? Thx

**Comment by kohlhase** [comment]
You should talk to Mihnea Iancu who is here.

**Comment by m-iancu** [comment]
@defeo We have a docker image for the MathHub system (odk instance is at http://odk.mathhub.info/) 
The github repo is at:  https://github.com/KWARC/mathhub_docker
The docker image is at: https://hub.docker.com/r/kwarc/mathhub/

**Comment by tkw1536** [comment]
For localmh the links are 
https://github.com/KWARC/localmh_docker (github repo) and
https://hub.docker.com/r/kwarc/localmh/ (docker image)

**Comment by embray** [comment]
Started typing of my status report and findings in the description.  Will add more on Monday.

**Comment by defeo** [comment]
Hi, @embray, are you done editing? If so, I will take over and add info on other containers.

**Comment by embray** [comment]
@defeo I'm done for now, thanks for asking.  See also #172 :)

**Comment by defeo** [comment]
I completed the description of the deliverable above. @fangohr, @m-iancu, @tkw1536, @alex-konovalov, please check that the descriptions of each project match what has been done so far.

**Comment by alex-konovalov** [comment]
@defeo thanks - done. We could add more details if need be. I haven't seen a working GAP Docker container on Windows yet, and the part on Sage has a very detailed explanation of the difficulties that Docker faces there. If this would work, that will be of an interest to many GAP users on Windows, unable to run packages that require compilation and are not Windows-compatible.

**Comment by defeo** [comment]
Thanks @alex-konovalov. I think this is enough. SageMath is very detailed, true, that's because @embray is still experimenting heavily. I'm optimistic on the Windows environment, and that's going to be a huge win for GAP too.

**Comment by embray** [comment]
@alex-konovalov To be clear, the Docker container for Sage includes GAP too.  If the desired the same installer I'm building for Sage, as well as the container build, could be used purely for GAP as well with the appropriate rebranding.  In fact GAP is easier since there's a working Ubuntu package for it.

**Comment by alex-konovalov** [comment]
@embray thanks for clarification. If I am not mistaken, the GAP in Sage Docker container is not 100% identical to the GAP distribution from the GAP website, though. It's interesting to see how we may incorporate a complete GAP install there.

**Comment by embray** [comment]
Just as a question on process--@defeo you closed this issue, which makes sense on one hand since the "deliverable" is more or less complete in this case.  But there's more work to keep doing on this as well.  So the question is then where best to continue tracking that work?

**Comment by embray** [comment]
For now I've moved the status report to a wiki page and updated the issue summary: https://github.com/OpenDreamKit/OpenDreamKit/wiki/D3.1-Virtual-images-and-containers

This should allow for more sane editing.

**Comment by defeo** [comment]
I closed the issue, because the deliverable is complete, and I submitted it to the EU portal. For now we can continue post here, or in the wiki.

Let's continue the discussion on the sane way to handle deliverables in #172. We can also wait for input from @nthiery.

**Comment by embray** [comment]
Note: I just updated the link in the description for the final report: it was previously broken.

Also, I am working on a brief presentation specifically on efforts in containerization for the status meeting in Bremen.  Is there anyone who has worked on the deliverable who has anything specific they'd like me to include?

I will reach out individually to some people if I don't hear anything here.

**Comment by alex-konovalov** [comment]
@embray thanks - I think that GAP Docker container (https://hub.docker.com/u/gapsystem/) may be relevant.

**Comment by embray** [comment]
@alex-konovalov Well yes, certainly that would be mentioned. But what about it?

**Comment by bpilorget** [comment]
# Deliverables update
@defeo  As you know , during the interim review in Bremen, reviewers gave advice to improve deliverables. UPSud tried to follow their advice by updating D5.1 #107
The main critic is that deliverables should be static document. Therefore we went through the #107 report to check if all links were still working and also if the information contained on the links was necessary for the understanding of the report.
As a result when it was necessary annexes were added to the #107 report.

Can you please do the same for this deliverable and add annexes when need be?

**Comment by defeo** [comment]
> Can you please do the same for this deliverable and add annexes when need be?

@bpilorget, I confirm that this deliverable is complete and needs no appendices. Do you want me to update the pdf with the new "disclaimer" on links being informative and so and so?

**Comment by bpilorget** [comment]
Yes that would be good, thanks.

**Comment by defeo** [comment]
> Yes that would be good, thanks.

Done

**Comment by nthiery** [comment]
Resubmitted on the EU portal (finally).
